src/system/random/basic.lean

```diff
@@ -0,0 +1,435 @@
+
+This module provides tools for formulating computations guided by randomness and for
+defining objects that can be created randomly.
+
+## Main definitions
+  * `rand` monad for computations guided by randomness;
+  * `random` class for objects that can be generated randomly;
+    * `random` to generate one object;
+    * `random_r` to generate one object inside a range;
+    * `random_series` to generate an infinite series of objects;
+    * `random_series_r` to generate an infinite series of objects inside a range;
+  * `io.mk_generator` to create a new random number generator;
+  * `io.run_rand` to run a randomized computation inside the `io` monad;
+  * `tactic.run_rand` to run a randomized computation inside the `tactic` monad
+
+## Tags
+
+random monad io
+
+## References
+
+  * Similar library in Haskell: https://hackage.haskell.org/package/MonadRandom
+
+-/
+
+open list io applicative
+
+universes u v w
+
+/-- A monad to generate random objects using the generator type `g` -/
+@[reducible]
+def rand_g (g : Type) := state (ulift g)
+/-- A monad to generate random objects using the generator type `std_gen` -/
+@[reducible]
+def rand := rand_g std_gen
+
+instance (g : Type) : liftable (rand_g.{u} g) (rand_g.{v} g) :=
+@state_t.liftable' _ _ _ _ _ (equiv.ulift.trans.{u u u u u} equiv.ulift.symm)
+
+open ulift (hiding inhabited)
+
+/-- Generate one more `ℕ` -/
+def random.next {g : Type} [random_gen g] : rand_g g ℕ :=
+⟨ prod.map id up ∘ random_gen.next ∘ down ⟩
+
+local infix ` .. `:41 := set.Icc
+
+open stream
+
+/-- `random α` gives us machinery to generate values of type `α` -/
+class random (α : Type u) extends preorder α :=
+(random : Π (g : Type) [random_gen g], rand_g g α)
+(random_r : Π g [random_gen g] (x y : α),
+              x ≤ y →
+              rand_g g (x .. y))
+
+/-- is 2^31; multiplying by it shifts a number left by 31 bits,
+dividing by it shifts it right by 31 bits -/
+def shift_31l : ℕ :=
+by apply_normed 2^31
+
+namespace random
+
+variables (α : Type u) [_root_.random α]
+variables {g : Type} [random_gen g]
+
+/-- Use a state monad to generate a stream through corecursion -/
+def corec_state {σ α} (cmd : state σ α) (s : σ) : stream α :=
+stream.corec prod.fst (cmd.run ∘ prod.snd) (cmd.run s)
+
+/-- create a new random number generator distinct from the one stored in the state -/
+def split : rand_g g g := ⟨ prod.map id up ∘ random_gen.split ∘ down ⟩
+
+/-- generate an infinite series of random values of type `α` -/
+def random_series : rand_g g (stream α) :=
+do gen ← liftable.up split,
+   pure $ corec_state (random α g) gen
+
+variables {α}
+
+/-- generate an infinite series of random values of type `α` between `x` and `y` -/
+def random_series_r (x y : α) (h : x ≤ y) : rand_g g (stream (x .. y)) :=
+do gen ← liftable.up split,
+   pure $ corec_state (random_r g x y h) gen
+
+end random
+
+namespace tactic.interactive
+
+/-- Some functions require a non-empty range as a parameter. This
```

Comment:
You're right, the reasoning is not very specific. I could call it `assumption_or_dec_trivial` but the name is hardly any simpler than the definition. And it would be subject to change if I need to add tricks to check ranges.